llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -19791,15 +19792,63 @@ static SDValue performIntrinsicCombine(SDNode *N,
           EVT::getVectorVT(*DAG.getContext(), PromVT.getVectorElementType(),
                            VT.getVectorElementCount());
 
-      Res = DAG.getNode(ISD::INTRINSIC_WO_CHAIN, DL, WhileVT, ID,
-                        N->getOperand(1), N->getOperand(2));
+      SDValue Res = DAG.getNode(ISD::INTRINSIC_WO_CHAIN, DL, WhileVT, ID,
+                                N->getOperand(1), N->getOperand(2));
       Res = DAG.getNode(ISD::SIGN_EXTEND, DL, PromVT, Res);
       Res = DAG.getNode(ISD::EXTRACT_SUBVECTOR, DL, ExtVT, Res,
                         DAG.getConstant(0, DL, MVT::i64));
       Res = DAG.getNode(ISD::TRUNCATE, DL, VT, Res);
+
+      return Res;
     }
-    return Res;
+
+    if (!Subtarget->hasSVE2p1() && !Subtarget->hasSME2())
+      return SDValue();
+
+    if (!N->hasNUsesOfValue(2, 0))
+      return SDValue();
+
+    auto It = N->use_begin();
+    SDNode *Lo = *It++;
```

Comment:
Could you explain, please? AFAIK, there is no guaranteed order of the uses of a value, so either part can appear first or second.

Reply:
That is part of DAG output.
  t6: nxv32i1 = llvm.get.active.lane.mask TargetConstant:i64<167>, t2, t4
    t8: nxv16i1 = extract_subvector t6, Constant:i64<0>
  t12: ch,glue = CopyToReg t0, Register:nxv16i1 $p0, t8
    t10: nxv16i1 = extract_subvector t6, Constant:i64<16>
AFAIU the first extract has offset 0 and the second  has offset 16. I believe it will always be in that order, as this probably will be a for loop starting from 0

Reply:
Then in this case, without making changes, the first extract will be lower half of the vector (`Lo`) and the second will be the higher part of the vector (`Hi`), which means that we will not do swaps, isn't it?

(but still will work correctly if for some reason the uses come in a different order).

Reply:
Do you mean that the order in the textual dump is usually the opposite to the order in `uses()` ?

Reply:
Ah, I see what you mean. However:
* there's no guarantee that we get the uses in any particular order, even if we generated them in some order. Hence we need the test and the swap (which is super cheap anyway, something like two conditional moves)
* we do generate the extracts in the "opposite" order, i.e. the last extract is the least significant bits, offset 0. This is done for another reason (so we lower the chance that a flag setting instruction will appear before the condition), but it has the fortunate side effect that the swap will likely not be needed, c.f. https://github.com/momchil-velikov/llvm-project/blob/766c78aebecfacb27dc1a8e221aebca37d93359d/llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp#L537